### PR TITLE
Add MailPublisher utility class

### PR DESCRIPTION
**Pull request by dnlkoch**
In this PR you will find a new utility class `MailPublisher` providing basic functionality being used for sending mails by the use of an existing mail-server. The configuration needed by Spring's `JavaMailSenderImpl` including a mail-template example is added to the webapp-archetype as well.

A real usage in your app (using the mail-template) would look like (pseudocode):

```java
public class MyMailClass {

    // autowire the mailPublisher class
    @Autowired
    private MailPublisher mailPublisher;

    // autowire the registrationMailMessageTemplate
    @Autowired
    private SimpleMailMessage registrationMailMessageTemplate;

    public void sendMail() {

        // set the mail recipient
        registrationMailMessageTemplate.setTo(User.getName());
        
        // fill the mail text template
        registrationMailMessageTemplate.setText(
                String.format(
                        registrationMailMessageTemplate.getText(),
                        User.getName(),
                        YourTokenizedRegistrationURL
                 )
        );

        // send the mail
        mailPublisher.sendMail(registrationMailMessageTemplate);
    }
}
```

### CODE REVIEW

**Comment by buehner** [comment]
Thanks for your work! :ok_hand:

I have some minor notes:

* Please try to remove the `spring-context-support` dependency from the security module (as it is included in util now)
* Please try to remove the `spring-test` dependency from the web, security and dao module (as it is included in util now)
* Rename `test-context.xml` to `mail-test-context.xml` (because there may be other test contexts in future)
* Try to use `@BeforeClass` and `@AfterClass` to start/stop GreenMail in the mail test. This should enhance the test performance as this would not run before each test, but only once before all tests.
* Please remove all that kagawa shinji and BVB stuff. It is simply annoying and provoces eye cancer... :wink:

File: src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+
+		registrationMailMessageTemplate.setTo(to);
+
+		registrationMailMessageTemplate.setText(
+				String.format(
+						registrationMailMessageTemplate.getText(),
+						str1,
+						str2
+				)
+		);
```
Comment (marcjansen) [inline src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java]: Indentation seems a bit off here.

File: src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+		// wait for max 5s for 1 email to arrive
+		// waitForIncomingEmail() is useful if you're sending stuff
+		// asynchronously in a separate thread
+		assertTrue(greenMail.waitForIncomingEmail(5000, 1));
+
+		// the BCC header isn't carried in the message-header, therefore we
+		// can't access it directly and we can only test if two mails were
+		// sent without having a TO-header set
+		Message[] messages = greenMail.getReceivedMessages();
```
Comment (marcjansen) [inline src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java]: Comment is misleading. Copy and paste?

File: src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
```
Comment (marcjansen) [inline src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java]: See above

> Comment is misleading. Copy and paste?

**Comment by marcjansen** [comment]
Very nice work, @dnlkoch!

I have minor comments which you may want to adress. In contrary to @buehner I have no problem with using `kagawa` and `shinji`. Tests tend to have a funny side note and should be fun to work with. I am unsure about whwther we can use the logo of that team (license). I would suggest to replace it with something that we may use without problems. What about this one: 

image

https://upload.wikimedia.org/wikipedia/commons/thumb/7/76/Minamoto_no_Yoritomo.jpg/188px-Minamoto_no_Yoritomo.jpg (it is public domain)

Thanks again, and please merge once the "valid" concerns of @buehner are resolved.

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/mail/MailPublisher.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+			LOG.error("Could not send the mail: " + e.getMessage());
+		}
+	}
+
+	/**
+	 * @param mailSender the mailSender to set
+	 */
+	public void setMailSender(JavaMailSender mailSender) {
+		this.mailSender = mailSender;
+	}
```
Comment (marcjansen) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/mail/MailPublisher.java]: Why do we only have a getter for the `defaultMailSender` and not for the `mailSender`?. I'd suggest add one for `mailSender`.

File: src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java (hunk 2 above)
Comment (dnlkoch) [inline src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java]: Yes, it's not mentioned to be here!

File: src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java (hunk 3 above)
Comment (dnlkoch) [inline src/shogun2-util/src/test/java/de/terrestris/shogun2/util/mail/MailPublisherTest.java]: Yes, it's not mentioned to be here!

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/mail/MailPublisher.java (hunk 4 above)
Comment (dnlkoch) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/mail/MailPublisher.java]: +1 
I added the getter method.

**Comment by dnlkoch** [comment]
Thanks @buehner and @marcjansen for your valuable notes!

**Comment by dnlkoch** [comment]
I think we can work with the logo as it is in the public-domain (see [here](https://commons.wikimedia.org/wiki/File:Borussia_Dortmund_logo.svg))

**Comment by marcjansen** [comment]
Good to go for me.

**Pull request merged**